core-bundle/src/Resources/contao/library/Contao/Environment.php

```diff
@@ -177,161 +174,78 @@ protected static function documentRoot()
 	 *
 	 * @return string The query string
 	 */
-	protected static function queryString()
+	protected static function queryString(Request $request)
 	{
-		if (!isset($_SERVER['QUERY_STRING']))
-		{
-			return '';
-		}
-
-		return static::encodeRequestString($_SERVER['QUERY_STRING']);
+		return static::encodeRequestString($request->getQueryString());
 	}
 
 	/**
 	 * Return the request URI [path]?[query] (e.g. /contao/index.php?id=2)
 	 *
 	 * @return string The request URI
 	 */
-	protected static function requestUri()
+	protected static function requestUri(Request $request)
 	{
-		if (!empty($_SERVER['REQUEST_URI']))
-		{
-			$arrComponents = parse_url($_SERVER['REQUEST_URI']);
-
-			if ($arrComponents === false)
-			{
-				$strRequest = $_SERVER['REQUEST_URI'];
-			}
-			else
-			{
-				$strRequest = $arrComponents['path'] . (isset($arrComponents['query']) ? '?' . $arrComponents['query'] : '');
-			}
-		}
-		else
-		{
-			$strRequest = '/' . preg_replace('/^\//', '', static::get('scriptName')) . (!empty($_SERVER['QUERY_STRING']) ? '?' . $_SERVER['QUERY_STRING'] : '');
-		}
-
-		return static::encodeRequestString($strRequest);
+		return static::encodeRequestString($request->getRequestUri());
 	}
 
 	/**
 	 * Return the first eight accepted languages as array
 	 *
 	 * @return array The languages array
 	 */
-	protected static function httpAcceptLanguage()
+	protected static function httpAcceptLanguage(Request $request)
 	{
-		$arrAccepted = array();
-		$arrLanguages = array();
-
-		// The implementation differs from the original implementation and also works with .jp browsers
-		preg_match_all('/([a-z]{1,8}(-[a-z]{1,8})?)\s*(;\s*q\s*=\s*(1|0\.[0-9]+))?/i', $_SERVER['HTTP_ACCEPT_LANGUAGE'], $arrAccepted);
-
-		// Remove all invalid locales
-		foreach ($arrAccepted[1] as $v)
-		{
-			$chunks = explode('-', $v);
-
-			// Language plus dialect, e.g. en-US or fr-FR
-			if (isset($chunks[1]))
-			{
-				$locale = $chunks[0] . '-' . strtoupper($chunks[1]);
-
-				if (preg_match('/^[a-z]{2}(-[A-Z]{2})?$/', $locale))
-				{
-					$arrLanguages[] = $locale;
-				}
-			}
-
-			$locale = $chunks[0];
-
-			// Language only, e.g. en or fr (see #29)
-			if (preg_match('/^[a-z]{2}$/', $locale))
-			{
-				$arrLanguages[] = $locale;
-			}
-		}
-
-		return \array_slice(array_unique($arrLanguages), 0, 8);
+		return str_replace('_', '-', $request->getLanguages());
```

Comment:
Previously, we only returned the first eight languages because the header can be manipulated to be a really long string. Should we restore this behavior?

Reply:
As the core never uses it, it might make sense to restore the original behavior for better BC :+1: I will update the PR

Reply:
Done in e8d61e853ba7348c90637044d17b097f129551b3